### DISCUSSION TITLE
How can I inject my own UserDetails or AbstractAuthenticationToken post Authentication?

### DISCUSSION THREAD

**Discussion by virtex-br** [ch4mpy/spring-addons#228] [Q&A] [answered]
I have a filter chain set up for API keys which does something like this:

```java
ApiKeyAuthenticationToken authentication = new ApiKeyAuthenticationToken(myDomainUserObject);
SecurityContextHolder.getContext().setAuthentication(authentication);
filterChain.doFilter(request, response);
``` 

I'd like to do something similar when I'm using the JwtAuthenticationToken provided by the spring-addons-starter-oidc plugin.

Is there a simple way to provide a UserDetails service or similar?

Thanks in advance

**Comment by ch4mpy** [comment, marked as answer]
With OAuth2, the "user-service" is the authorization server: that's it which is responsible for providing user data as OAuth2 claims (inside a JWT access token or as introspection response payload for resource servers, and as ID token payload or userinfo response payload for clients). On resource servers and clients, you shouldn't be looping to the DB to get user details & authorities and, more importantly, you should not duplicate this authorization server data in your resource server's DB.

With `oauth2ResourceServer` and a JWT decoder, the `Authentication` instance is returned by a `Converter<Jwt, AbstractAuthenticationToken>`. The default implementation is `JwtAuthenticationConverter` which builds a `JwtAuthenticationToken`. It can be replaced with `http.oauth2ResourceServer(oauth2 -> oauth2.jwt(jwt -> jwt.jwtAuthenticationConverter(...)))`, which configures the default authentication manager.

**As stated in [the doc](https://github.com/ch4mpy/spring-addons/tree/master/spring-addons-starter-oidc#113-authentication-converter)**, `spring-addons-starter-oidc` exposes a default `Converter<Jwt, AbstractAuthenticationToken>` as a `@ConditionalOnMissingBean` and uses it to configure the authentication manager. So **all you have to do is exposing a `Converter<Jwt, AbstractAuthenticationToken>` as a bean** (a `@Bean`, `@Componenent`, or alike). As a bonus, if using `@WithJwt` in your tests, this bean will be picked by the test authentication factory and you'll have your custom `Authentication` implementation in tests too.

Note that `spring-addons-starter-oidc` exposes a default authorities converter too, and it is pretty convenient. So you can ignore or override it, but the best option is probably to inject the default authorities converter in your authentication converter factory.

You can find demos of that in the [resource-server_with_oauthentication](https://github.com/ch4mpy/spring-addons/blob/593055050927c85db4b390a250e23242f6b334a5/samples/tutorials/resource-server_with_oauthentication/src/main/java/com/c4soft/springaddons/tutorials/ResourceServerWithOAuthenticationApplication.java#L45) and [resource-server_with_specialized_oauthentication](https://github.com/ch4mpy/spring-addons/blob/593055050927c85db4b390a250e23242f6b334a5/samples/tutorials/resource-server_with_specialized_oauthentication/src/main/java/com/c4soft/springaddons/tutorials/SecurityConfig.java#L24) tutorials.

As a side note, you might find the `OAuthentication<OpenIdClaimSet>` interesting as a base-class for your custom `Authentication` implementation, as done [in the second tutorial](https://github.com/ch4mpy/spring-addons/blob/593055050927c85db4b390a250e23242f6b334a5/samples/tutorials/resource-server_with_specialized_oauthentication/src/main/java/com/c4soft/springaddons/tutorials/ProxiesAuthentication.java#L16C44-L16C76) (have a look at the API with your IDE auto-completion).

**Comment by virtex-br** [comment]
As always, a fantastic answer - thank you
